File: app/src/main/java/com/lkms/data/repository/implement/java/AuthRepositoryImpl.java
package com.lkms.data.repository.implement.java;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.lkms.data.dal.SupabaseClient;
import com.lkms.data.model.User;
import com.lkms.data.repository.IAuthRepository;

import java.lang.reflect.Type;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.List;
import java.util.Scanner;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class AuthRepositoryImpl implements IAuthRepository {

    private static final String SUPABASE_URL = "https://your-project.supabase.co/rest/v1"; // chỉnh lại URL thật
    private static final String SUPABASE_API_KEY = "your-api-key"; // thêm key nếu cần
    private static final Gson gson = new Gson();

    private final ExecutorService executor = Executors.newFixedThreadPool(4);

    @Override
    public void login(String email, String password, AuthCallback callback) {
        if (email == null || email.isBlank() || password == null || password.isBlank()) {
            if (callback != null) callback.onError("Email and Password is not blanked.");
            return;
        }

        CompletableFuture.runAsync(() -> {
            try {
                // Gọi API REST của Supabase (dạng PostgREST)
                String endpoint = SUPABASE_URL + "/User?select=*&email=eq." + email + "&password=eq." + password;

                String json = getJson(endpoint);

                Type listType = new TypeToken<List<User>>() {}.getType();
                List<User> users = gson.fromJson(json, listType);

                if (users != null && !users.isEmpty()) {
                    if (callback != null) callback.onSuccess(users.get(0));
                } else {
                    if (callback != null) callback.onError("Email or Password is wrong.");
                }
            } catch (Exception e) {
                if (callback != null) callback.onError("Login failed: " + e.getMessage());
            }
        }, executor);
    }

    @Override
    public void logout(LogoutCallback callback) {
    }

    // Hàm gọi GET Supabase (thay thế cho coroutine client trong Kotlin)
    private String getJson(String endpoint) throws Exception {
        URL url = new URL(endpoint);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");
        conn.setRequestProperty("apikey", SUPABASE_API_KEY);
        conn.setRequestProperty("Authorization", "Bearer " + SUPABASE_API_KEY);
        conn.setRequestProperty("Accept", "application/json");

        int responseCode = conn.getResponseCode();
        if (responseCode != HttpURLConnection.HTTP_OK) {
            throw new RuntimeException("HTTP error code: " + responseCode);
        }

        try (Scanner scanner = new Scanner(conn.getInputStream())) {
            scanner.useDelimiter("\\A");
            return scanner.hasNext() ? scanner.next() : "";
        }
    }
}

File: app/src/main/java/com/lkms/data/repository/implement/AuthRepositoryImpl.kt
//package com.lkms.data.repository.implement
//
//import android.content.Context
//import com.lkms.data.dal.SupabaseClient
//import com.lkms.data.repository.IAuthRepository
//import kotlinx.coroutines.CoroutineScope
//import kotlinx.coroutines.Dispatchers
//import io.github.jan.supabase.postgrest.postgrest
//import com.lkms.data.model.kotlin.User
//import kotlinx.coroutines.*
//
//class AuthRepositoryImpl : IAuthRepository {
//    private val client = SupabaseClient.client
//    private val scope = CoroutineScope(Dispatchers.IO)
//
//    override fun login(
//        email: String?,
//        password: String?,
//        callback: IAuthRepository.AuthCallback?
//    ) {
//        if (email.isNullOrBlank() || password.isNullOrBlank()) {
//            callback?.onError("Email and Password is not blanked.")
//            return
//        }
//
//        scope.launch {
//            try {
//                val users = client.postgrest["User"]
//                    .select {
//                        filter {
//                            eq("email", email)
//                            eq("password", password)
//                        }
//                    }
//                    .decodeList<User>()
//
//                if (users.isNotEmpty()) {
//                    callback?.onSuccess(users.first())
//                } else {
//                    callback?.onError("Email or Password is wrong.")
//                }
//            } catch (e: Exception) {
//                callback?.onError("Login failed: ${e.message}")
//            }
//        }
//    }
//
//    override fun logout(callback: IAuthRepository.LogoutCallback?) {
//    }
//
//}